Rename meta-data to tags

toolmantim opened the pull request:
There has been a long proposed change to the Buildkite Agent terminology and command line API for rename "meta-data" to "tags".

This would change:

```bash
buildkite-agent start \
  --meta-data docker \
  --meta-data postgres=9 \
  --meta-data postgres=9.4 \
  --meta-data queue=runner \
  --meta-data queue=docker
```

to:

```bash
buildkite-agent start \
  --tag docker \
  --tag postgres=9 \
  --tag postgres=9.4 \
  --tag queue=runner \
  --tag queue=docker
```

- [x] First cut at the code
- [x] General feedback
- [x] Show we add `queue` as it's own command line arg? - answer: maybe, but let's move to a different PR
- [x] Should we add support for comma or space separation, for easier setting of multiple values via an env var? e.g. `export BUILDKITE_AGENT_TAG=docker,postgres=9,postgres=9.4` - answer: it works already
- [x] Do we want/need to version bump the agent API? - answer: nope
- [x] Should we add backwards compatibility? Do we support existing config files, env vars, and/or command line args? - answer: yes
- [x] How does this map to the `pipeline.yml` agent targeting rules? - answer: no changes
- [x] Second cut at the code
- [x] Add backwards compatibility

toolmantim commented:
Another thing to consider is the design on the other side, actually running a job that targets the agent once you've got it running.

Current `pipeline.yml` syntax:

```yml
steps:
  - command: docker run my-app-image script/test.sh
    agents:
      queue: docker
      docker: true
      postgres: 9
```

Ideas:

```yml
steps:
  - command: docker run hello-world
    queue: deploy
    tags:
      docker: true
      postgres: 9
```

```yml
steps:
  - command: docker run hello-world
    agents:
      queue: deploy
      tags:
        docker: true
        postgres: 9
```

Do we support lists/arrays?

sj26 commented on agent/agent_pool.go:
```diff
@@ -14,16 +14,16 @@ import (
+	Priority       string
+	Tags           []string
+	EC2            bool
+	EC2Tags        bool
```
I'd stick with a prefix here. `EC2` is tags and `EC2Tags` is EC2 Tags as tags. So maybe `Tags`, `TagsFromEC2`, `TagsFromEC2Tags` and `TagsFromGCP` or something?

toolmantim replied:
Oh yeah, good idea!

sj26 commented:
Long overdue. Let's make it happen!

keithpitt commented:
This is rad!

> Should we add support for comma or space separation, for easier setting of multiple values via an env var? e.g. export BUILDKITE_AGENT_TAG=docker,postgres=9,postgres=9.4

I do believe that should just work already!

> Do we want/need to version bump the agent API?

I'd just leave the stuff in `api/` as is, then we don't need to worry about bumping the Agent API at the moment.

> Should we add backwards compatibility? Do we support existing config files, env vars, and/or command line args?

I definitely think it needs to be backwards compatible with the old. Ship it with V3, and we can deprecate it in 3.1 or something. We'd probably want to throw out some `buildkite-warning` type lines when they try and use the old syntax.

> Move queue to it's own command line arg

Do you think making them 2 seperate things (`--tag foo` and `--queue blah`) makes it a little confusing as to their relationship? Queues are tags but tags are not queues... When do you want to use a tag? Should I use a tag? Should I use a queue? This probably is a stupid question, but maybe it's better to just keep them all as tags? `--tag queue=blah` - at least it's a little shorter..

In my head I imagine it working a little bit like EC2 "Name" tags. They're still a tag, you define them as a tag, but the key has some magic behind it.

That way we don't need to change any other bits around the edges. We wouldn't need to change the `agents` pipeline.yml syntax either...

I know I've suggested the `queue` split before, but took me actually seeing it to question whether or not it was a good idea :P

What ya think?

toolmantim commented:
Thanks @keithpitt!

You're right that `BUILDKITE_AGENT_META_DATA=docker,postgres=9,postgres=9.4` works AOK already! 🎉 

> Do you think making them 2 seperate things (--tag foo and --queue blah) makes it a little confusing as to their relationship? Queues are tags but tags are not queues... When do you want to use a tag? Should I use a tag? Should I use a queue? 

I think you just described exactly why I wanted them as separate things. Because people need to stop and ask themselves those questions you just asked.

Tags and Queues aren't the same… you certainly don't want to treat 'queue' like it's any old tag… and I think we want them as separate things in the docs. People will want to, and need to, understand the difference.

Queues and tags are both core concepts for designing stuff with Buildkite. There's big conceptual and behavioural differences between the two. We're often recommending using a queue with a single agent to solve a bunch of modelling problems around deploys and such, a setup that tags can't really help with.

That we've got code that goes are parses tags in a whole number of places (`agent.tags.select =~ /\Aqueue=/`) is also maybe a good hint that it should be promoted to it's own 'thing'?

We can also make it backwards compatible with "tags", and not change the pipeline.yml syntax for now either. And I think that could work really well. So it should be relatively easy to keep as a cli change for now.

But in saying all that, I'm also okay with just changing `meta-data` → `tags` in this PR, and leaving any queue change stuff to another PR?

toolmantim commented:
Have updated the TODOs. Will get the code finished up and tested shortly too.

keithpitt commented:
> But in saying all that, I'm also okay with just changing meta-data → tags in this PR, and leaving any queue change stuff to another PR?

Yeah, I think that's a good idea. Moving to `tags` is good and I wouldn't want that discussion to get in the way of shipping the first rename!

toolmantim commented:
@sj26 @keithpitt in 1777a2752645577ab9f2b4bdae439efa7b6b7983 I've attempted to add support for loading in the old config values, so we don't break people's 3.0 beta agents… and instead we remove every major version upgrade.

```
$ buildkite-agent-dev start --token 'xxx' --meta-data moo=co
2017-04-27 16:32:33 NOTICE The config option `meta-data` has been renamed to `tags` and will be removed from the next version of Buildkite Agent

  _           _ _     _ _    _ _                                _
 | |         (_) |   | | |  (_) |                              | |
 | |__  _   _ _| | __| | | ___| |_ ___    __ _  __ _  ___ _ __ | |_
 | '_ \| | | | | |/ _` | |/ / | __/ _ \  / _` |/ _` |/ _ \ '_ \| __|
 | |_) | |_| | | | (_| |   <| | ||  __/ | (_| | (_| |  __/ | | | |_
 |_.__/ \__,_|_|_|\__,_|_|\_\_|\__\___|  \__,_|\__, |\___|_| |_|\__|
                                                __/ |
 http://buildkite.com/agent                    |___/

2017-04-27 16:32:33 NOTICE Starting buildkite-agent v3.0-beta.22 with PID: 95580
2017-04-27 16:32:33 NOTICE The agent source code can be found here: https://github.com/buildkite/agent
2017-04-27 16:32:33 NOTICE For questions and support, email us at: hello@buildkite.com
2017-04-27 16:32:33 INFO   Configuration loaded from: /usr/local/etc/buildkite-agent/buildkite-agent.cfg
2017-04-27 16:32:33 INFO   Registering agent with Buildkite...
2017-04-27 16:32:34 INFO   Successfully registered agent "tim.local-1" with tags []
2017-04-27 16:32:34 INFO   Connecting to Buildkite...
2017-04-27 16:32:35 INFO   Agent successfully connected
2017-04-27 16:32:35 INFO   You can press Ctrl-C to stop the agent
2017-04-27 16:32:35 INFO   Waiting for work...
```

The only thing that stumped me is I can't get the value to be passed through. You can see in the above output, tags is `[]`. Any ideas? My golang foo is weak.

toolmantim commented:
Ok, this should be ready to go!

If you use the old args you get a warning, but it still works AOK:

```
$ buildkite-agent-dev start --meta-data foo=bar --token xxx
2017-04-28 18:13:02 WARN   The config option `meta-data` has been renamed to `tags`. Please update your configuration.

  _           _ _     _ _    _ _                                _
 | |         (_) |   | | |  (_) |                              | |
 | |__  _   _ _| | __| | | ___| |_ ___    __ _  __ _  ___ _ __ | |_
 | '_ \| | | | | |/ _` | |/ / | __/ _ \  / _` |/ _` |/ _ \ '_ \| __|
 | |_) | |_| | | | (_| |   <| | ||  __/ | (_| | (_| |  __/ | | | |_
 |_.__/ \__,_|_|_|\__,_|_|\_\_|\__\___|  \__,_|\__, |\___|_| |_|\__|
                                                __/ |
 http://buildkite.com/agent                    |___/

2017-04-28 18:13:02 NOTICE Starting buildkite-agent v3.0-beta.22 with PID: 82423
2017-04-28 18:13:02 NOTICE The agent source code can be found here: https://github.com/buildkite/agent
2017-04-28 18:13:02 NOTICE For questions and support, email us at: hello@buildkite.com
2017-04-28 18:13:02 INFO   Configuration loaded from: $HOME/.buildkite-agent/buildkite-agent.cfg
2017-04-28 18:13:02 INFO   Registering agent with Buildkite...
2017-04-28 18:13:03 INFO   Successfully registered agent "Tims-iMac.local-2" with tags [foo=bar]
2017-04-28 18:13:03 INFO   Connecting to Buildkite...
2017-04-28 18:13:05 INFO   Agent successfully connected
2017-04-28 18:13:05 INFO   You can press Ctrl-C to stop the agent
2017-04-28 18:13:05 INFO   Waiting for work...
^C2017-04-28 18:13:05 INFO   Gracefully stopping agent. Since there is no job running, the agent will disconnect immediately
2017-04-28 18:13:05 INFO   Disconnecting Tims-iMac.local-2...
```

If you use both config options you get the same deprecation warning, but then it errors with a message that shows both the config names and their values so you can go searching for where it might be defined (cfg file, environment variable, or command line arg):

```
$ buildkite-agent-dev start --tags foo=bar --meta-data foo2=bar2 --token xxx
2017-04-28 18:17:38 WARN   The config option `meta-data` has been renamed to `tags`. Please update your configuration.
2017-04-28 18:17:38 FATAL  Can't set config option `meta-data=[foo2=bar2]` because `tags=[foo=bar]` has already been set
exit status 1
```

toolmantim commented:
I was hoping we could hide these from the cli `--help` output, but `Hidden` doesn't seem to work in `urfave/cli` so instead I just added some usage text:

```
$ buildkite-agent-dev start --help
Usage:

   buildkite-agent start [arguments...]

Description:

   When a job is ready to run it will call the "bootstrap-script"
   and pass it all the environment variables required for the job to run.
   This script is responsible for checking out the code, and running the
   actual build script defined in the pipeline.

   The agent will run any jobs within a PTY (pseudo terminal) if available.

Example:

   $ buildkite-agent start --token xxx

Options:

   --config value                               Path to a configuration file [$BUILDKITE_AGENT_CONFIG]
   --token value                                Your account agent token [$BUILDKITE_AGENT_TOKEN]
   --name value                                 The name of the agent [$BUILDKITE_AGENT_NAME]
   --priority value                             The priority of the agent (higher priorities are assigned work first) [$BUILDKITE_AGENT_PRIORITY]
   --disconnect-after-job                       Disconnect the agent after running a job [$BUILDKITE_AGENT_DISCONNECT_AFTER_JOB]
   --disconnect-after-job-timeout value         When --disconnect-after-job is specified, the number of seconds to wait for a job before shutting down (default: 120) [$BUILDKITE_AGENT_DISCONNECT_AFTER_JOB_TIMEOUT]
   --tags value                                 Tags for the agent (default is "queue=default") [$BUILDKITE_AGENT_TAGS]
   --tags-from-ec2                              Include the host's EC2 meta-data as tags (instance-id, instance-type, and ami-id) as meta-data [$BUILDKITE_AGENT_TAGS_FROM_EC2]
   --tags-from-ec2-tags                         Include the host's EC2 tags as tags [$BUILDKITE_AGENT_TAGS_FROM_EC2_TAGS]
   --tags-from-gcp                              Include the host's Google Cloud meta-data as tags (instance-id, machine-type, preemptible, project-id, region, and zone) [$BUILDKITE_AGENT_TAGS_FROM_GCP]
   --git-clone-flags value                      Flags to pass to the "git clone" command (default: "-v") [$BUILDKITE_GIT_CLONE_FLAGS]
   --git-clean-flags value                      Flags to pass to "git clean" command (default: "-fxdq") [$BUILDKITE_GIT_CLEAN_FLAGS]
   --bootstrap-script value                     Path to the bootstrap script (default: "buildkite-agent bootstrap") [$BUILDKITE_BOOTSTRAP_SCRIPT_PATH]
   --build-path value                           Path to where the builds will run from [$BUILDKITE_BUILD_PATH]
   --hooks-path value                           Directory where the hook scripts are found [$BUILDKITE_HOOKS_PATH]
   --plugins-path value                         Directory where the plugins are saved to [$BUILDKITE_PLUGINS_PATH]
   --no-pty                                     Do not run jobs within a pseudo terminal [$BUILDKITE_NO_PTY]
   --no-automatic-ssh-fingerprint-verification  Don't automatically verify SSH fingerprints [$BUILDKITE_NO_AUTOMATIC_SSH_FINGERPRINT_VERIFICATION]
   --no-command-eval                            Don't allow this agent to run arbitrary console commands [$BUILDKITE_NO_COMMAND_EVAL]
   --experiment value                           Enable experimental features within the buildkite-agent [$BUILDKITE_AGENT_EXPERIMENT]
   --endpoint value                             The Agent API endpoint (default: "https://agent.buildkite.com/v3") [$BUILDKITE_AGENT_ENDPOINT]
   --no-color                                   Don't show colors in logging [$BUILDKITE_AGENT_NO_COLOR]
   --debug                                      Enable debug mode [$BUILDKITE_AGENT_DEBUG]
   --debug-http                                 Enable HTTP debug mode, which dumps all request and response bodies to the log [$BUILDKITE_AGENT_DEBUG_HTTP]
   --meta-data value                            Deprecated. Renamed to --tags [$BUILDKITE_AGENT_META_DATA]
   --meta-data-ec2                              Deprecated. Renamed to --tags-from-ec2 [$BUILDKITE_AGENT_META_DATA_EC2]
   --meta-data-ec2-tags                         Deprecated. Renamed to --tags-from-ec2-tags [$BUILDKITE_AGENT_TAGS_FROM_EC2_TAGS]
   --meta-data-gcp                              Deprecated. Renamed to --tags-from-gcp [$BUILDKITE_AGENT_META_DATA_GCP]
```

keithpitt commented:
@toolmantim I just remembered, it could be this code here that's still showing `Hidden` options:

keithpitt commented:
Doh! Pressed enter too quickly...

https://github.com/buildkite/agent/blob/master/main.go#L28

We have a custom CLI display that may need to be tweaked somehow...

toolmantim commented:
@keithpitt oh, that'd be it! I'll update that template to use `.VisibleFlags` instead of `.Flags`

toolmantim commented:
Alright, I've made this change and the fields are hidden ⚡️

I was wondering if we should rename `tags` to `tag`, for the case of:

```
buildkite-agent start \
  --tag foo \
  --tag bar=baz
```

But as I was writing the cli flag description I wanted to tell people how it works with the environment variable version too.

So instead of `tag` and mentioning that you can specify the flag multiple times, I just kept it as `tags` and changed the description to `A comma-separated list of tags for the agent (e.g. \"linux\" or \"mac,xcode=8\")`. This means it works the same in the agent config file, env vars, and via cli flags. That example also shows people that the value is optional.

I chose not to mention that you can specify the cli flag multiple times and they'll be merged, but maybe we should?

Here's the current full help output:

```
$ buildkite-agent-dev start --help
Usage:

   buildkite-agent start [arguments...]

Description:

   When a job is ready to run it will call the "bootstrap-script"
   and pass it all the environment variables required for the job to run.
   This script is responsible for checking out the code, and running the
   actual build script defined in the pipeline.

   The agent will run any jobs within a PTY (pseudo terminal) if available.

Example:

   $ buildkite-agent start --token xxx

Options:

   --config value                               Path to a configuration file [$BUILDKITE_AGENT_CONFIG]
   --token value                                Your account agent token [$BUILDKITE_AGENT_TOKEN]
   --name value                                 The name of the agent [$BUILDKITE_AGENT_NAME]
   --priority value                             The priority of the agent (higher priorities are assigned work first) [$BUILDKITE_AGENT_PRIORITY]
   --disconnect-after-job                       Disconnect the agent after running a job [$BUILDKITE_AGENT_DISCONNECT_AFTER_JOB]
   --disconnect-after-job-timeout value         When --disconnect-after-job is specified, the number of seconds to wait for a job before shutting down (default: 120) [$BUILDKITE_AGENT_DISCONNECT_AFTER_JOB_TIMEOUT]
   --tags value                                 A comma-separated list of tags for the agent (e.g. "linux" or "mac,xcode=8") [$BUILDKITE_AGENT_TAGS]
   --tags-from-ec2                              Include the host's EC2 meta-data as tags (instance-id, instance-type, and ami-id) [$BUILDKITE_AGENT_TAGS_FROM_EC2]
   --tags-from-ec2-tags                         Include the host's EC2 tags as tags [$BUILDKITE_AGENT_TAGS_FROM_EC2_TAGS]
   --tags-from-gcp                              Include the host's Google Cloud meta-data as tags (instance-id, machine-type, preemptible, project-id, region, and zone) [$BUILDKITE_AGENT_TAGS_FROM_GCP]
   --git-clone-flags value                      Flags to pass to the "git clone" command (default: "-v") [$BUILDKITE_GIT_CLONE_FLAGS]
   --git-clean-flags value                      Flags to pass to "git clean" command (default: "-fxdq") [$BUILDKITE_GIT_CLEAN_FLAGS]
   --bootstrap-script value                     Path to the bootstrap script (default: "buildkite-agent bootstrap") [$BUILDKITE_BOOTSTRAP_SCRIPT_PATH]
   --build-path value                           Path to where the builds will run from [$BUILDKITE_BUILD_PATH]
   --hooks-path value                           Directory where the hook scripts are found [$BUILDKITE_HOOKS_PATH]
   --plugins-path value                         Directory where the plugins are saved to [$BUILDKITE_PLUGINS_PATH]
   --no-pty                                     Do not run jobs within a pseudo terminal [$BUILDKITE_NO_PTY]
   --no-automatic-ssh-fingerprint-verification  Don't automatically verify SSH fingerprints [$BUILDKITE_NO_AUTOMATIC_SSH_FINGERPRINT_VERIFICATION]
   --no-command-eval                            Don't allow this agent to run arbitrary console commands [$BUILDKITE_NO_COMMAND_EVAL]
   --experiment value                           Enable experimental features within the buildkite-agent [$BUILDKITE_AGENT_EXPERIMENT]
   --endpoint value                             The Agent API endpoint (default: "https://agent.buildkite.com/v3") [$BUILDKITE_AGENT_ENDPOINT]
   --no-color                                   Don't show colors in logging [$BUILDKITE_AGENT_NO_COLOR]
   --debug                                      Enable debug mode [$BUILDKITE_AGENT_DEBUG]
   --debug-http                                 Enable HTTP debug mode, which dumps all request and response bodies to the log [$BUILDKITE_AGENT_DEBUG_HTTP]
```

toolmantim commented:
‼️🐴
